tools/state-viewer/src/cli.rs

```diff
@@ -606,3 +620,214 @@ impl ViewTrieCmd {
+            };
+            println!("shard id    {shard_id:#?} benchmark finished");
+        }
+    }
+
+    fn get_trie(
+        &self,
+        shard_id: ShardId,
+        shard_layout: &near_primitives::shard_layout::ShardLayout,
+        chunk_header: &near_primitives::sharding::ShardChunkHeader,
+        store: &Store,
+    ) -> Trie {
+        let shard_uid = ShardUId::from_shard_id_and_layout(shard_id, shard_layout);
+        // Note: here we get the previous state root but the shard layout
+        // corresponds to the current epoch id. In practice shouldn't
+        // matter as the shard layout doesn't change.
+        let state_root = chunk_header.prev_state_root();
+        // Question for reviewer - there are different TrieStorage
+        // implementation, which would be best suited here? I picked this
+        // one since it looked simplest and I don't think there is much
+        // point in using caches.
+        let storage = TrieDBStorage::new(store.clone(), shard_uid);
+        // Question for reviewer - my assumption is that flat storgage
+        // should be disabled or is irrelevant for this task, can you
+        // confirm please?
+        let flat_storage_chunk_view = None;
+        Trie::new(Rc::new(storage), state_root, flat_storage_chunk_view)
+    }
+
+    fn iter_trie(&self, trie: &Trie, prune_condition: Option<Box<dyn Fn(&Vec<u8>) -> bool>>) {
+        let start = Instant::now();
+        let mut node_count = 0;
+        let mut error_count = 0;
+        let iter = trie.iter_with_prune_condition(prune_condition);
+        let iter = match iter {
+            Ok(iter) => iter,
+            Err(err) => {
+                println!("iter error {err:#?}");
+                return;
+            }
+        };
+        for item in iter {
+            node_count += 1;
+            if let Some(limit) = self.limit {
+                if limit < node_count {
+                    break;
+                }
+            }
+
+            let (key, value) = match item {
+                Ok((key, value)) => (key, value),
+                Err(err) => {
+                    println!("Failed to iterate node with error: {err}");
+                    error_count += 1;
+                    continue;
+                }
+            };
+
+            if self.print {
+                let state_record = StateRecord::from_raw_key_value(key.clone(), value);
+                Self::print_state_record(&state_record);
+            }
+        }
+        let duration = start.elapsed();
+        println!("node count  {node_count}");
+        println!("error count {error_count}");
+        println!("time        {duration:?}");
+    }
+
+    fn shallow_iter_prune_condition(key_nibbles: &Vec<u8>) -> bool {
+        // Need at least 2 nibbles for the column type byte.
+        if key_nibbles.len() < 2 {
+            return false;
+        }
+        // Keys for interesting nodes have even length
+        if key_nibbles.len() % 2 != 0 {
```

Comment:
Hmm, this would directly violate the condition you've written down that if a prefix is pruned then the key is also pruned, right?

Should we just ignore the last nibble in this computation if there are an odd number of nibbles?